src/Workspaces/Core/Portable/Serialization/SerializableSourceText.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Serialization
+{
+    internal sealed class SerializableSourceText
+    {
+        public ITemporaryTextStorageWithName? Storage { get; }
+        public SourceText? Text { get; }
+
+        public SerializableSourceText(ITemporaryTextStorageWithName storage)
+            : this(storage, text: null)
+        {
+        }
+
+        public SerializableSourceText(SourceText text)
+            : this(storage: null, text)
+        {
+        }
+
+        private SerializableSourceText(ITemporaryTextStorageWithName? storage, SourceText? text)
+        {
+            Debug.Assert(storage is not null || text is not null);
```

Comment:
```suggestion
            Debug.Assert(storage is not null || text is not null);
            Debug.Assert(storage is null || text is null);
```

Reply:
or
```suggestion
            Debug.Assert(storage is null != text is null);
```

Reply:
It's allowed to have both, as long as they reference the same source.

Reply:
how can you construct such a thing?  all i see are two public constructors, each which only take one value, but pass null for the other.  It seems like in all cases, it's one or the other.

Reply:
➡️ Now documented